File: go/problem_53.go
package main

type StackQueue struct {
	inbox, outbox []int
}

func (q *StackQueue) enqueue(val int) {
	q.inbox = append(q.inbox, val)
}

func (q *StackQueue) dequeue() (int, bool) {
	if len(q.outbox) == 0 {
		for len(q.inbox) > 0 {
			n := len(q.inbox)
			q.outbox = append(q.outbox, q.inbox[n-1])
			q.inbox = q.inbox[:n-1]
		}
	}
	if len(q.outbox) == 0 {
		return 0, false
	}
	n := len(q.outbox)
	v := q.outbox[n-1]
	q.outbox = q.outbox[:n-1]
	return v, true
}

func problem53() {
	println("Problem 53:")
	q := &StackQueue{}
	q.enqueue(1); q.enqueue(2); q.enqueue(3)
	v, _ := q.dequeue(); assertEqual(v, 1)
	v, _ = q.dequeue(); assertEqual(v, 2)
	q.enqueue(4)
	v, _ = q.dequeue(); assertEqual(v, 3)
	v, _ = q.dequeue(); assertEqual(v, 4)
	_, ok := q.dequeue()
	if ok {
		panic("should be empty")
	}
	println("All Tests Passed")
}

File: lua/problem_53.lua
local M = {}

function M.new_stack_queue()
    local q = {inbox = {}, outbox = {}}
    function q:enqueue(v) self.inbox[#self.inbox+1] = v end
    function q:dequeue()
        if #self.outbox == 0 then
            while #self.inbox > 0 do
                self.outbox[#self.outbox+1] = table.remove(self.inbox)
            end
        end
        if #self.outbox == 0 then return nil end
        return table.remove(self.outbox)
    end
    return q
end

function M.problem53()
    print("Problem 53:")
    local q = M.new_stack_queue()
    q:enqueue(1); q:enqueue(2); q:enqueue(3)
    assert(q:dequeue() == 1)
    assert(q:dequeue() == 2)
    q:enqueue(4)
    assert(q:dequeue() == 3)
    assert(q:dequeue() == 4)
    assert(q:dequeue() == nil)
    print("All Tests Passed")
end

return M
